src/toast/templates/hwpss.py

```diff
@@ -0,0 +1,534 @@
+# Copyright (c) 2023-2024 by the parties listed in the AUTHORS file.
+# All rights reserved.  Use of this source code is governed by
+# a BSD-style license that can be found in the LICENSE file.
+
+import ast
+import json
+from collections import OrderedDict
+
+import h5py
+import numpy as np
+from scipy.linalg import lu_factor, lu_solve
+
+from ..data import Data
+from ..mpi import MPI
+from ..observation import default_values as defaults
+from ..timing import function_timer
+from ..traits import Bool, Float, Instance, Int, Unicode, trait_docs
+from ..utils import Logger
+from ..vis import set_matplotlib_backend
+from .amplitudes import Amplitudes
+from .template import Template
+
+
+@trait_docs
+class Hwpss(Template):
+    """This template represents the HWP synchronous signal."""
+
+    # Notes:  The TraitConfig base class defines a "name" attribute.  The Template
+    # class (derived from TraitConfig) defines the following traits already:
+    #    data             : The Data instance we are working with
+    #    view             : The timestream view we are using
+    #    det_data         : The detector data key with the timestreams
+    #    det_data_units   : The units of the detector data
+    #    det_mask         : Bitmask for per-detector flagging
+    #    det_flags        : Optional detector solver flags
+    #    det_flag_mask    : Bit mask for detector solver flags
+    #
+
+    hwp_angle = Unicode(
+        defaults.hwp_angle, allow_none=True, help="Observation shared key for HWP angle"
+    )
+
+    hwp_flags = Unicode(
+        defaults.shared_flags,
+        allow_none=True,
+        help="Observation shared key for HWP flags",
+    )
+
+    hwp_flag_mask = Int(
+        defaults.det_mask_invalid,
+        help="Bit mask to use when considering valid HWP angle values.",
+    )
+
+    harmonics = Int(9, help="Number of harmonics to consider in the expansion")
+
+    times = Unicode(defaults.times, help="Observation shared key for timestamps")
+
+    debug_plots = Unicode(
+        None,
+        allow_none=True,
+        help="If not None, make debugging plots in this directory",
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def _initialize(self, new_data):
+        log = Logger.get()
+
+        # For each harmonic with the sin, cos, and time drift terms for each.
```

Comment:
In the case where the time drift in the data is zero, I don't think the extra parameters will cause a problem since there is so much data being used to constrain a small number of amplitudes.  It would require a bit of work to change the code to support this, so would prefer to punt on that unless it is needed.